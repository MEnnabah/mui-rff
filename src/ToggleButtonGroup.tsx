import React, { ReactNode } from 'react';

import {
	FormHelperTextProps,
	ToggleButtonGroup as MuiToggleButtonGroup,
	ToggleButtonGroupProps as MuiToggleButtonGroupProps,
} from '@mui/material';

import { ErrorMessage, ShowErrorFunc, showErrorOnChange, useFieldForErrors } from './Util';
import { Field } from 'react-final-form';

export interface ToggleButtonGroupProps extends Partial<Omit<MuiToggleButtonGroupProps, 'onChange'>> {
	name: string;
	helperText?: string;
	formHelperTextProps?: Partial<FormHelperTextProps>;
	showError?: ShowErrorFunc;
	children?: ReactNode;
}

export function ToggleButtonGroup(props: ToggleButtonGroupProps) {
	const {
		name,
		helperText,
		formHelperTextProps,
		children,
		fullWidth = true,
		showError = showErrorOnChange,
		...restToggleButtonGroupProps
	} = props;

	const field = useFieldForErrors(name);
	const isError = showError(field);

	return (
		<Field name={name}>
			{(fieldProps) => {
				return (
					<>
						<MuiToggleButtonGroup
							value={fieldProps.input.value}
							fullWidth={fullWidth}
							disabled={props.disabled}
							onChange={(_, value) => {
								fieldProps.input.onChange(value);
							}}
							{...restToggleButtonGroupProps}
						>
							{children}
						</MuiToggleButtonGroup>
						<ErrorMessage
							showError={isError}
							meta={field.meta}
							formHelperTextProps={formHelperTextProps}
							helperText={helperText}
						/>
					</>
				);
			}}
		</Field>
	);
}
