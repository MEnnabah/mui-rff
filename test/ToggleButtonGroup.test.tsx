import * as Yup from 'yup';
import { Button, ToggleButton } from '@mui/material';
import { Form } from 'react-final-form';
import { ToggleButtonGroup, makeValidateSync } from '../src';
import { fireEvent, render } from '@testing-library/react';
import React from 'react';

interface ComponentProps {
	initialValues: FormData;
	validator?: any;
}

interface FormData {
	selection: string[];
}

describe('ToggleButtonGroup', () => {
	const initialValues: FormData = {
		selection: ['A'],
	};

	function ToggleButtonGroupComponent({ initialValues, validator }: ComponentProps) {
		const onSubmit = (values: FormData) => {
			console.log(values);
		};
		return (
			<Form
				onSubmit={onSubmit}
				initialValues={initialValues}
				validate={validator}
				render={({ handleSubmit, submitting }) => (
					<form onSubmit={handleSubmit} noValidate>
						<ToggleButtonGroup name="toggle">
							<ToggleButton value={'A'}>
								<p>Option A</p>
							</ToggleButton>
						</ToggleButtonGroup>
						<Button
							variant="contained"
							color="primary"
							type="submit"
							disabled={submitting}
							data-testid="submit"
						>
							Submit
						</Button>
					</form>
				)}
			/>
		);
	}

	it('renders without errors', async () => {
		const rendered = render(<ToggleButtonGroupComponent initialValues={initialValues} />);
		expect(rendered).toMatchSnapshot();
	});

	it('turns red if empty and required', async () => {
		const validateSchema = makeValidateSync(
			Yup.object().shape({
				selection: Yup.array(Yup.string().required()).min(1),
			}),
		);

		const rendered = render(
			<ToggleButtonGroupComponent initialValues={{ selection: [] }} validator={validateSchema} />,
		);

		const submit = await rendered.findByTestId('submit');
		fireEvent.click(submit);

		expect(rendered).toMatchSnapshot();
	});
});
