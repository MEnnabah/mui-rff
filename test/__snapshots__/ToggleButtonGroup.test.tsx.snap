// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButtonGroup renders without errors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        novalidate=""
      >
        <div
          class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-mgs850-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="A"
          >
            <p>
              Option A
            </p>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      novalidate=""
    >
      <div
        class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-mgs850-MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="A"
        >
          <p>
            Option A
          </p>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="submit"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ToggleButtonGroup turns red if empty and required 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        novalidate=""
      >
        <div
          class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-mgs850-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="A"
          >
            <p>
              Option A
            </p>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      novalidate=""
    >
      <div
        class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-mgs850-MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="A"
        >
          <p>
            Option A
          </p>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="submit"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
